### PR TITLE
Julia & Victoria -- Carets

### PR DESCRIPTION
**Pull request by vsawchuk**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Explain how you came up with the design of your ERD, based on the seed data. | We used the customer and movie fields we had currently available, and knew we would like to create a join table to contain rentals since it would otherwise require a many-to-many relationship between customers and movies. |
| Describe a set of positive and negative test cases you implemented for a model.  | A rental could be created with valid data, and could not be created with invalid data. |
| Describe a set of positive and negative test cases you implemented for a controller.  | The movies controller can find a movie with a valid id, and will return a 404 status code if the movie id was invalid. |
| How does your API respond when bad data is sent to it? | Typically we send back a 400 or 404 response code depending on the situation. |
| Describe one of your custom model methods and why you chose to wrap that functionality into a method. | We added the checkout_count as an instance method in the Customer model because we didn't want to store this logic in the controller nor as a field in our customers table in Postgres. This reduces storage requirements for our database. |
| Do you have any recommendations on how we could improve this project for the next cohort? |  |
| Link to Trello | We don't have one ¯\_(ツ)_/¯ (ps this emoji is mostly joking, but we just followed the directions in order)|
| Link to ERD | Customers -|--< Rentals >--|- Movies 
julia_victoria_erd
|

### CODE REVIEW

File: test/models/customer_test.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+require "test_helper"
+require "awesome_print"
+
+describe Customer do
+
+  it "a customer object can be saved to the database" do
```
Comment (CheezItMan) [inline test/models/customer_test.rb:6]: So no validations for Customers?

File: app/models/customer.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class Customer < ApplicationRecord
+  has_many :rentals
+
+  def self.index_customers
```
Comment (CheezItMan) [inline app/models/customer.rb:4]: Interesting method to create the JSON, why not use a serializer?

File: test/models/movie_test.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+require "test_helper"
+
+describe Movie do
+
+  it "a movie object can be saved to the database" do
```
Comment (CheezItMan) [inline test/models/movie_test.rb:5]: negative tests?  What about when the inventory, or available inventory is 0.  Can you create a movie with negative inventory?

File: test/controllers/movies_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    end
+
+    it "can create a new movie" do
+      proc {post create_movie_path, params: {title: "Raiders of the Lost Ark"}}.must_change "Movie.count", 1
+      must_respond_with :ok
+      body = JSON.parse(response.body)
+      body["id"].must_be_instance_of Integer
+    end
+
+    it "cannot save a duplicate movie (same title)" do
```
Comment (CheezItMan) [inline test/controllers/movies_controller_test.rb:95]: What about if the value posted is missing some needed fields?

**Comment by CheezItMan** [comment]
# Video Store
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | Good number of commits, good commit messages, both team members contributed |
| Comprehension questions	| Check |
| **General** |  |
| Business Logic in Models | Good number of method in the models dealing with checkout, inventory etc. |
| All 3 required endpoints return expected JSON data | Check |
| Requests respond with appropriate HTTP Status Code  | Check |
| Errors are reported | Check |
|  **Testing** |  |
| Passes all Smoke Tests | Yep, plus the extras |
| Model Tests - all relations, validations, and custom functions test positive & negative cases | You're missing some negative tests and tests for edge cases, left some notes in your code.  |
| Controller Tests - URI parameters and data in the request body have positive & negative cases | Overall good, left some notes in your code.  |
| **Optionals** |  |
| POST routes use URI parameter and request body to add a new record to the database | Well done! |
| GET /customers shows how many movies are checked out by a customer | Check! |
| GET /movies/:id shows updated inventory | Check |
|  **Overall** | Well done, some minor issues with your testing, but very nice work especially on the extras. |
